docs/design/architecture_3.0/README.md

```diff
@@ -0,0 +1,145 @@
+|                            | Route               | Stage 1               |
+|                            | `MacVtap`           | Stage 3               |
+| Storage                    | Virtio-fs           | Stage 1               |
+|                            | `nydus`             | Stage 2               |
+| Hypervisor                 | `Dragonball`        | Stage 1               |
+|                            | QEMU                | Stage 2               |
+|                            | ACRN                | Stage 3               |
+|                            | Cloud Hypervisor    | Stage 3               |
+|                            | Firecracker         | Stage 3               |
+
+## FAQ
+
+- Are the "service", "message dispatcher" and "runtime handler" all part of the single Kata 3.x runtime binary?
+
+  <font size=2>Yes. They are components in Kata 3.x runtime. </font><br>
+  <font size=2>1. Service is an interface, which is responsible for handling multiple services like task service, image service and etc. </font><br>
+  <font size=2>2. Message dispatcher, it is used to match multiple requests from the service module. </font><br>
+  <font size=2>3. Runtime handler is used to deal with the operation for sandbox and container. </font>
+- What is the name of the Kata 3.x runtime binary?
+  
+  <font size=2>containerd-shim-kata -v2</font>
+
+- Is the Kata 3.x design compatible with the containerd shimv2 architecture? 
+  
+  <font size=2>Yes. It is designed to follow the functionality of go version kata.</font>
+
+- How will users migrate to the Kata 3.x architecture?
+  
+  <font size=2>The migration plan will be provided before the Kata 3.x is merging into the main branch.</font>
+
+- Is `Dragonball` limited to its own built-in VMM? Can the `Dragonball` system be configured to work using an external `Dragonball` VMM/hypervisor?
+
+  <font size=2>The `Dragonball` could work as an external hypervisor. However, stability and performance is challenging in this case. Built in VMM could optimise the container overhead, and it's easy to maintain stability.</font>
+- QEMU, Cloud Hypervisor and Firecracker support are planned, but how that would work. Are they working in separate process?
+ 
+  <font size=2>Yes. They are unable to work as built in VMM.</font>
+
+- What is `upcall`?
+  
+   <font size=2> `Dbs-upcall` is a `vsock-based` direct communication tool between VMM and guests. The server side of the `upcall` is a driver in guest kernel (kernel patches are needed for this feature) and it'll start to serve the requests once the kernel has started. And the client side is in VMM , it'll be a thread that communicates with VSOCK through `uds`. We have accomplished device hotplug / hot-unplug directly through `upcall` in order to avoid virtualization of ACPI  to minimize virtual machine's overhead. And there could be many other usage through this direct communication channel.<br> It's already open source.</font><br>
```

Comment:
Can you provide details of this driver and the kernel patches. Also, some more specific questions:

- Are the patches going to be upstreamed?
- Has the process of upstreaming started? Is there an ETA for when they might land in Linus' tree?
- Are these patches platform-specific or would they work for any architecture that supports VSOCK?
- Could the kernel driver be replaced with a userland daemon in the guest using loopback VSOCK?
- The URL below says the kernel patches work with 4.19, but do they also work with 5.15+ (aka the current Kata kernel version)?
- The fact that upcall allows comms between the VMM and the guest suggests that this architecture might be incompatible with https://github.com/confidential-containers where the VMM should have no knowledge of what happens inside the VM.

/cc @mythi, @amshinde, @sameo.

Reply:
> Can you provide details of this driver and the kernel patches. Also, some more specific questions:
I will try my best to give some background about this driver. Dragonball is derived from firecracker, so it relies on virtio-mmio instead of PCI/ACPI etc. The upcall is used to hotplug CPU/memory/MMIO devices, and solves two issues:
1) avoid dependency on PCI/ACPI
2) avoid dependency on udevd within guest and get deterministic results for hotplug operations
So upcall is an alternative to ACPI based CPU/memory/device hotplug. And we may cooperate with the community to add support for ACPI based CPU/memory/device hotplug if needed.
 
> * Are the patches going to be upstreamed?
We are planning to open-source this driver.
We plan to submit patches of this driver to kata.

> * Has the process of upstreaming started? Is there an ETA for when they might land in Linus' tree?
No concrete plan to upstream this to Linus' tree yet.

> * Are these patches platform-specific or would they work for any architecture that supports VSOCK?
It's almost platform independent, but some message related to CPU hotplug are platform dependent.

> * Could the kernel driver be replaced with a userland daemon in the guest using loopback VSOCK?
We need to create device nodes for hot-added CPU/memory/devices, so it's not easy for userspace daemon to do these tasks.

> * The URL below says the kernel patches work with 4.19, but do they also work with 5.15+ (aka the current Kata kernel version)?
Forward compatibility should be achievable, we have ported it to 5.10 based kernel.

> * The fact that upcall allows comms between the VMM and the guest suggests that this architecture might be incompatible with https://github.com/confidential-containers where the VMM should have no knowledge of what happens inside the VM.
Per my understanding it's acceptable due to:
1) TDX doesn't support CPU/memory hotplug yet.
2) For ACPI based device hotplug, it depends on ACPI DSDT table, and the guest kernel will execute ASL code to handle during handling those hotplug event. And it should be easier to audit vsock based communication than ACPI ASL methods:)

> 
> /cc @mythi, @amshinde, @sameo.